datafusion/expr/src/expr.rs

```diff
@@ -270,6 +270,27 @@ pub enum GroupingSet {
     GroupingSets(Vec<Vec<Expr>>),
 }
 
+impl GroupingSet {
+    /// Return all distinct exprs in the grouping set. For `CUBE` and `ROLLUP` this
+    /// is just the underlying list of exprs. For `GROUPING SET` we need to deduplicate
+    /// the exprs in the underlying sets.
+    pub fn distinct_expr(&self) -> Vec<Expr> {
+        match self {
+            GroupingSet::Rollup(exprs) => exprs.clone(),
+            GroupingSet::Cube(exprs) => exprs.clone(),
+            GroupingSet::GroupingSets(groups) => {
+                let mut exprs: Vec<Expr> = vec![];
+                for exp in groups.iter().flatten() {
+                    if !exprs.contains(exp) {
```

Comment:
This is N^2 in the number of grouping sets -- probably not an issue, I just figured I would point it out

Reply:
Yeah, this is unfortunate